// @insforge/sdk/dist/ssr/middleware.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/ssr/middleware.ts
var middleware_exports = {};
__export(middleware_exports, {
  DEFAULT_ACCESS_TOKEN_COOKIE: () => DEFAULT_ACCESS_TOKEN_COOKIE,
  DEFAULT_REFRESH_TOKEN_COOKIE: () => DEFAULT_REFRESH_TOKEN_COOKIE,
  clearAuthCookies: () => clearAuthCookies,
  getAccessTokenCookieName: () => getAccessTokenCookieName,
  getRefreshTokenCookieName: () => getRefreshTokenCookieName,
  setAuthCookies: () => setAuthCookies,
  updateSession: () => updateSession
});
module.exports = __toCommonJS(middleware_exports);

// src/lib/jwt.ts
function decodeBase64Url(input) {
  const normalized = input.replace(/-/g, "+").replace(/_/g, "/");
  const padded = normalized.padEnd(
    normalized.length + (4 - normalized.length % 4) % 4,
    "="
  );
  const binary = atob(padded);
  const bytes = Uint8Array.from(binary, (char) => char.charCodeAt(0));
  return new TextDecoder().decode(bytes);
}
function getJwtExpiration(token) {
  if (!token) return null;
  const [, payload] = token.split(".");
  if (!payload) return null;
  try {
    const parsed = JSON.parse(decodeBase64Url(payload));
    if (typeof parsed.exp !== "number" || !Number.isFinite(parsed.exp)) {
      return null;
    }
    return new Date(parsed.exp * 1e3);
  } catch {
    return null;
  }
}
function isJwtExpiredOrExpiring(token, leewaySeconds = 60) {
  if (!token) return false;
  const expires = getJwtExpiration(token);
  if (!expires) return true;
  return expires.getTime() <= Date.now() + leewaySeconds * 1e3;
}

// src/ssr/cookies.ts
var DEFAULT_ACCESS_TOKEN_COOKIE = "insforge_access_token";
var DEFAULT_REFRESH_TOKEN_COOKIE = "insforge_refresh_token";
var EXPIRED_DATE = /* @__PURE__ */ new Date(0);
function getAccessTokenCookieName(names) {
  return names?.accessToken ?? DEFAULT_ACCESS_TOKEN_COOKIE;
}
function getRefreshTokenCookieName(names) {
  return names?.refreshToken ?? DEFAULT_REFRESH_TOKEN_COOKIE;
}
function getCookieValue(cookies, name) {
  if (!cookies) return null;
  const value = cookies.get(name);
  if (typeof value === "string") return value || null;
  if (value && typeof value.value === "string") return value.value || null;
  return null;
}
function getCookieValueFromHeader(cookieHeader, name) {
  if (!cookieHeader) return null;
  const parts = cookieHeader.split(";");
  for (const part of parts) {
    const [rawName, ...rawValue] = part.trim().split("=");
    if (rawName !== name) continue;
    try {
      return decodeURIComponent(rawValue.join("="));
    } catch {
      return rawValue.join("=");
    }
  }
  return null;
}
function defaultCookieOptions() {
  const secure = typeof process !== "undefined" ? process.env.NODE_ENV === "production" : typeof location !== "undefined" && location.protocol === "https:";
  return {
    path: "/",
    sameSite: "lax",
    secure
  };
}
function accessTokenCookieOptions(token, overrides) {
  return {
    ...defaultCookieOptions(),
    httpOnly: false,
    expires: getJwtExpiration(token) ?? void 0,
    ...overrides
  };
}
function refreshTokenCookieOptions(token, overrides) {
  return {
    ...defaultCookieOptions(),
    httpOnly: true,
    expires: getJwtExpiration(token) ?? void 0,
    ...overrides
  };
}
function expiredCookieOptions(overrides) {
  const { expires: _expires, maxAge: _maxAge, ...safeOverrides } = overrides ?? {};
  return {
    ...defaultCookieOptions(),
    ...safeOverrides,
    expires: EXPIRED_DATE,
    maxAge: 0
  };
}
function setCookie(cookies, name, value, options) {
  if (!cookies?.set) return;
  cookies.set(name, value, options);
}
function deleteCookie(cookies, name, options) {
  if (!cookies) return;
  if (cookies.set) {
    cookies.set(name, "", expiredCookieOptions(options));
    return;
  }
  cookies.delete?.(name);
}
function serializeCookie(name, value, options = {}) {
  const parts = [`${encodeURIComponent(name)}=${encodeURIComponent(value)}`];
  if (options.maxAge !== void 0) parts.push(`Max-Age=${options.maxAge}`);
  if (options.domain) parts.push(`Domain=${options.domain}`);
  if (options.path) parts.push(`Path=${options.path}`);
  if (options.expires) parts.push(`Expires=${options.expires.toUTCString()}`);
  if (options.httpOnly) parts.push("HttpOnly");
  if (options.secure) parts.push("Secure");
  if (options.sameSite) {
    const sameSite = options.sameSite.charAt(0).toUpperCase() + options.sameSite.slice(1);
    parts.push(`SameSite=${sameSite}`);
  }
  return parts.join("; ");
}
function appendSetCookie(headers, name, value, options) {
  headers.append("Set-Cookie", serializeCookie(name, value, options));
}
function setAuthCookies(cookies, tokens, settings = {}) {
  const accessName = getAccessTokenCookieName(settings.names);
  const refreshName = getRefreshTokenCookieName(settings.names);
  const accessOptions = accessTokenCookieOptions(
    tokens.accessToken,
    settings.options?.accessToken
  );
  setCookie(cookies, accessName, tokens.accessToken, accessOptions);
  if (tokens.refreshToken) {
    setCookie(
      cookies,
      refreshName,
      tokens.refreshToken,
      refreshTokenCookieOptions(
        tokens.refreshToken,
        settings.options?.refreshToken
      )
    );
  }
}
function clearAuthCookies(cookies, settings = {}) {
  const accessName = getAccessTokenCookieName(settings.names);
  const refreshName = getRefreshTokenCookieName(settings.names);
  const accessOptions = expiredCookieOptions(settings.options?.accessToken);
  const refreshOptions = expiredCookieOptions(settings.options?.refreshToken);
  deleteCookie(cookies, accessName, accessOptions);
  deleteCookie(cookies, refreshName, refreshOptions);
}
function setAuthCookieHeaders(headers, tokens, settings = {}) {
  const accessName = getAccessTokenCookieName(settings.names);
  const refreshName = getRefreshTokenCookieName(settings.names);
  appendSetCookie(
    headers,
    accessName,
    tokens.accessToken,
    accessTokenCookieOptions(tokens.accessToken, settings.options?.accessToken)
  );
  if (tokens.refreshToken) {
    appendSetCookie(
      headers,
      refreshName,
      tokens.refreshToken,
      refreshTokenCookieOptions(
        tokens.refreshToken,
        settings.options?.refreshToken
      )
    );
  }
}
function clearAuthCookieHeaders(headers, settings = {}) {
  appendSetCookie(
    headers,
    getAccessTokenCookieName(settings.names),
    "",
    expiredCookieOptions(settings.options?.accessToken)
  );
  appendSetCookie(
    headers,
    getRefreshTokenCookieName(settings.names),
    "",
    expiredCookieOptions(settings.options?.refreshToken)
  );
}

// src/types.ts
var InsForgeError = class _InsForgeError extends Error {
  constructor(message, statusCode, error, nextActions) {
    super(message);
    this.name = "InsForgeError";
    this.statusCode = statusCode;
    this.error = error;
    this.nextActions = nextActions;
  }
  static fromApiError(apiError) {
    return new _InsForgeError(
      apiError.message,
      apiError.statusCode,
      apiError.error,
      apiError.nextActions
    );
  }
};

// src/ssr/refresh.ts
var import_shared_schemas = require("@insforge/shared-schemas");
function jsonResponse(body, init = {}, headers = new Headers(init.headers)) {
  headers.set("Content-Type", "application/json");
  return new Response(JSON.stringify(body), {
    ...init,
    headers
  });
}
function normalizeError(error) {
  if (error instanceof InsForgeError) return error;
  if (error && typeof error === "object") {
    const body = error;
    return new InsForgeError(
      typeof body.message === "string" ? body.message : "Failed to refresh auth session",
      typeof body.statusCode === "number" ? body.statusCode : 500,
      typeof body.error === "string" ? body.error : import_shared_schemas.ERROR_CODES.UNKNOWN_ERROR
    );
  }
  return new InsForgeError(
    error instanceof Error ? error.message : "Failed to refresh auth session",
    500,
    import_shared_schemas.ERROR_CODES.UNKNOWN_ERROR
  );
}
async function readJson(response) {
  const contentType = response.headers.get("content-type");
  if (!contentType?.includes("json")) return null;
  return response.json();
}
function readRefreshToken(options) {
  if (options.refreshToken) return options.refreshToken;
  const refreshCookieName = getRefreshTokenCookieName(options.names);
  const cookieValue = getCookieValue(options.cookies, refreshCookieName);
  if (cookieValue) return cookieValue;
  return getCookieValueFromHeader(
    options.request?.headers.get("cookie"),
    refreshCookieName
  );
}
async function refreshAuth(options = {}) {
  const headers = new Headers();
  const refreshToken = readRefreshToken(options);
  if (!refreshToken) {
    clearAuthCookieHeaders(headers, options);
    const error2 = new InsForgeError(
      "Refresh token cookie is missing",
      401,
      import_shared_schemas.ERROR_CODES.AUTH_UNAUTHORIZED
    );
    return {
      response: jsonResponse(
        {
          error: error2.error,
          message: error2.message,
          statusCode: error2.statusCode
        },
        { status: error2.statusCode },
        headers
      ),
      data: null,
      accessToken: null,
      refreshToken: null,
      error: error2
    };
  }
  let { baseUrl, anonKey } = options;
  try {
    baseUrl || (baseUrl = process.env.NEXT_PUBLIC_INSFORGE_URL);
    anonKey || (anonKey = process.env.NEXT_PUBLIC_INSFORGE_ANON_KEY);
  } catch {
  }
  if (!baseUrl || !anonKey) {
    throw new Error(
      "Missing InsForge baseUrl or anonKey. Pass baseUrl and anonKey to refreshAuth() or set NEXT_PUBLIC_INSFORGE_URL and NEXT_PUBLIC_INSFORGE_ANON_KEY."
    );
  }
  const fetchImpl = options.fetch ?? (globalThis.fetch ? globalThis.fetch.bind(globalThis) : void 0);
  if (!fetchImpl) {
    throw new Error(
      "Fetch is not available. Please provide a fetch implementation."
    );
  }
  const requestHeaders = new Headers(options.headers);
  requestHeaders.set("Authorization", `Bearer ${anonKey}`);
  requestHeaders.set("Content-Type", "application/json");
  requestHeaders.set("Accept", "application/json");
  let data = null;
  let error = null;
  try {
    const response = await fetchImpl(
      new URL("/api/auth/refresh?client_type=mobile", baseUrl).toString(),
      {
        method: "POST",
        headers: requestHeaders,
        body: JSON.stringify({ refresh_token: refreshToken })
      }
    );
    const body = await readJson(response);
    if (!response.ok) {
      error = normalizeError(
        body ?? {
          message: "Failed to refresh auth session",
          statusCode: response.status,
          error: import_shared_schemas.ERROR_CODES.UNKNOWN_ERROR
        }
      );
    } else {
      data = body;
    }
  } catch (caught) {
    error = normalizeError(caught);
  }
  if (error || !data?.accessToken) {
    clearAuthCookieHeaders(headers, options);
    const normalized = normalizeError(error);
    return {
      response: jsonResponse(
        {
          error: normalized.error,
          message: normalized.message,
          statusCode: normalized.statusCode
        },
        { status: normalized.statusCode || 500 },
        headers
      ),
      data: null,
      accessToken: null,
      refreshToken: null,
      error: normalized
    };
  }
  const nextRefreshToken = data.refreshToken ?? refreshToken;
  setAuthCookieHeaders(
    headers,
    {
      accessToken: data.accessToken,
      refreshToken: nextRefreshToken
    },
    options
  );
  const responseBody = {
    accessToken: data.accessToken,
    user: data.user,
    csrfToken: data.csrfToken
  };
  return {
    response: jsonResponse(responseBody, { status: 200 }, headers),
    data: responseBody,
    accessToken: data.accessToken,
    refreshToken: nextRefreshToken,
    error: null
  };
}

// src/ssr/update-session.ts
async function updateSession(options) {
  const accessCookieName = getAccessTokenCookieName(options.names);
  const refreshCookieName = getRefreshTokenCookieName(options.names);
  const accessToken = getCookieValue(
    options.requestCookies,
    accessCookieName
  );
  if (accessToken && !isJwtExpiredOrExpiring(accessToken, options.refreshLeewaySeconds)) {
    return {
      refreshed: false,
      accessToken,
      error: null
    };
  }
  const refreshToken = getCookieValue(
    options.requestCookies,
    refreshCookieName
  );
  if (!refreshToken) {
    if (accessToken) {
      clearAuthCookies(options.requestCookies, options);
      clearAuthCookies(options.responseCookies, options);
    }
    return {
      refreshed: false,
      accessToken: null,
      error: null
    };
  }
  const result = await refreshAuth({
    ...options,
    refreshToken
  });
  if (result.error || !result.accessToken) {
    clearAuthCookies(options.requestCookies, options);
    clearAuthCookies(options.responseCookies, options);
    return {
      refreshed: false,
      accessToken: null,
      error: result.error
    };
  }
  const tokens = {
    accessToken: result.accessToken,
    refreshToken: result.refreshToken ?? refreshToken
  };
  setAuthCookies(options.requestCookies, tokens, options);
  setAuthCookies(options.responseCookies, tokens, options);
  return {
    refreshed: true,
    accessToken: result.accessToken,
    error: null
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  DEFAULT_ACCESS_TOKEN_COOKIE,
  DEFAULT_REFRESH_TOKEN_COOKIE,
  clearAuthCookies,
  getAccessTokenCookieName,
  getRefreshTokenCookieName,
  setAuthCookies,
  updateSession
});
